docs: add agentstack guide (#2875)
neondatabase/website: content/guides/agentstack-neon.md

---
title: Building AI Agents with AgentStack and Neon
subtitle: Build a Web scraper AI Agent in minutes with AgentStack, Neon, and Firecrawl
author: dhanush-reddy
enableTableOfContents: true
createdAt: '2025-02-04T00:00:00.000Z'
updatedOn: '2025-02-04T00:00:00.000Z'
---

> As the world of AI agents rapidly evolves, the challenge lies in building and deploying them quickly and efficiently. Imagine creating intelligent agents that can not only perform complex tasks but also seamlessly interact with your data infrastructure, all without getting bogged down in complex code.

Reviewer: ```suggestion
The rapid evolution of AI agents has created a key challenge: how to build and deploy agents quickly and efficiently. Imagine creating intelligent agents that can not only perform complex tasks but also interact easily with your data infrastructure, without adding unnecessary complexity to the code.```

Reviewer: Suggested revision to avoid cliches and colloquialisms.

> This guide introduces [**AgentStack**](https://docs.agentstack.sh/introduction), a rapid development framework for AI agents, and **Neon**, the serverless Postgres database, to streamline the creation of powerful, data-driven AI agents. We will show you how to build a **Web Scraper AI Agent** in minutes, using AgentStack's intuitive CLI and tool integrations.

Reviewer: ```suggestion
This guide introduces [**AgentStack**](https://docs.agentstack.sh/introduction), a rapid development framework for AI agents, and **Neon**, the serverless Postgres database, to simplify the development of powerful, database-connected AI agents. We'll walk through building a **Web Scraper AI Agent** using AgentStack's CLI and tool integrations.
```

Reviewer: Modified the tone a bit. Also "data-driven" could imply data analysis/decision making, but here it's really about ability to interact with the database.

Author: modified as:

This guide introduces [**AgentStack**](https://docs.agentstack.sh/introduction), a rapid development framework, and Neon, the serverless Postgres database, and shows how they can be used together to build powerful AI agents with database integration. We'll walk through building a **Web Scraper AI Agent** using AgentStack's CLI and tool integrations.

> Forget tedious setup and manual coding! With AgentStack and Neon, you can generate complete agent workflows, create new agents and tasks with simple commands, and effortlessly integrate them with your data layer. Let's dive in and see how you can build intelligent agents faster than ever before.

Reviewer: ```suggestion
With AgentStack and Neon, you can generate complete agent workflows, create new agents and tasks with simple commands, and integrate them directly with your data layer. Let's get started.
```

## What you will build

This example will show you how to:

> - Set up an **AgentStack** project to create an AI Agent.

Reviewer: ```suggestion
- Set up an **AgentStack** project.
```

- Utilize the **AgentStack CLI** to generate agents and tasks.

> - Equip your agents with **tools** like **Neon** and **Firecrawl** for data storage and web scraping.

Reviewer: ```suggestion
- Equip your agents with **tools** like **Neon** for data storage and **Firecrawl** for web scraping.
```

- **uv package installer (or Poetry):** Recommended for faster dependency installation. Install `uv` as described here: [astral-sh/uv](https://github.com/astral-sh/uv?tab=readme-ov-file#installation). Alternatively, use `poetry`.
- **AgentStack CLI:** Install the AgentStack Command Line Interface (CLI). Follow the [Getting started with AgentStack](https://docs.agentstack.sh/installation) guide.
- **Accounts and API Keys:** You will need accounts and API keys for these services:
  - **OpenAI API key**: We will use OpenAI's `gpt-4o-mini` model to power AI agents. Get an OpenAI API key at [platform.openai.com](https://platform.openai.com).
  - **Neon account**: Sign up for a free Neon account at [neon.tech](https://console.neon.tech/signup). You will need a Neon API key to connect to your Neon database.
  - **Firecrawl account**: Sign up for a Firecrawl account at [firecrawl.dev](https://firecrawl.dev). You will need a Firecrawl API key to use the web scraping tool.
  - **AgentOps account**: Sign up for an AgentOps account at [agentops.ai](https://agentops.ai) to leverage agent observability features. You will need an AgentOps API key.

> ## Building the Web Scraper Agent

Reviewer: ```suggestion
## Building the Web Scraper agents
```

Author: the example is only for a Web scraper agent, so it can be singular?

I do agree Web scraper agent is composed of micro agents thus making plural, but on a whole it is acting a "Web scraper AI agent" in singular tone?

Reviewer: Same as above.

> Let's now build the Web Scraper AI agent using AgentStack, Neon, and Firecrawl. We will walk through each step, from initializing the project to running the agent crew.

Author: the example is only for a Web scraper agent, we can then make it to: "Let's start building our agent using AgentStack"?

> Let's now build the Web Scraper AI agent using AgentStack, Neon, and Firecrawl. We will walk through each step, from initializing the project to running the agent crew.

Reviewer: Sure, that works. I do think of the overall more as an agent project, with multiple agents making up the "crew". But this is good too.

- **`retrieve_web_crawl`**: Since `web_crawl` is an asynchronous operation, we use the `retrieve_web_crawl` action to get the results of a crawl that was initiated previously using the `web_crawl` action. This action requires the `crawl_id` that was returned by the initial `web_crawl` action. It checks the status of the crawl and returns the scraped content once the crawl is complete. Agents can use this action in a loop or after a delay to check for and retrieve crawl results, allowing for more complex workflows where web crawling is part of a longer process.

These actions equip our agents with powerful web scraping capabilities, ranging from simple single-page content extraction to comprehensive crawling of website sections, enabling them to gather web data effectively as part of their tasks.

> We can improve the efficiency of our agents by specifying their tool usage in crew.py. Since the `web_scraper` agent only needs firecrawl, the `content_storer` agent only needs neon, and the data_extractor agent needs no tool, assigning these tools directly in the configuration will reduce agent context memory and enable more focused task execution, mirroring human specialization.

Reviewer: nice!

- Load agent and task configurations.
- Instantiate the agent crew defined in `src/crew.py`.
- Execute the tasks in sequence.
- Utilize the `neon` and `firecrawl` tools within the agents' tasks as defined in `src/crew.py`.
- Print the final output to your terminal.

You should see the agent's execution logs and the final output, including the final SQL query generated by the `content_storer` agent.

![Web Scraper Agent Output 1](/docs/guides/agentstack-neon-example-output-1.png)
![Web Scraper Agent Output 2](/docs/guides/agentstack-neon-example-output-2.png)

### Verifying the output

After the agent run completes, check your terminal for the output. It should display the SQL query generated by the `content_storer` agent.

You can verify that the data has been stored in your Neon database by:

- Logging into your Neon account at [console.neon.tech](https://console.neon.tech).
- Navigating to your project and database.
- Clicking on the `Tables` tab to view the `posts` table created by the agent.

![Neon SQL Editor](/docs/guides/agentstack-neon-database-data.png)

<Admonition type="info" title="AgentOps: Gain Observability into your AI Agents">

> AgentStack seamlessly integrates with [**AgentOps**](https://www.agentops.ai) by default to provide crucial observability for your AI agent development. AgentOps, from the same team behind AgentStack, allows you to:

Reviewer: ```suggestion
AgentStack integrates with [**AgentOps**](https://www.agentops.ai) by default to provide crucial observability for your AI agent development. AgentOps, from the same team behind AgentStack, allows you to:
```

- **Visualize agent execution:** See step-by-step execution graphs of your agents, making it easy to understand workflows and debug issues.
- **Track LLM costs:** Monitor your spending on LLM providers to manage costs effectively.
- **Benchmark Agent performance:** Evaluate agent performance and ensure quality.
- **Enhance security:** Detect potential vulnerabilities like prompt injection.

![AgentOps Visualization](/docs/guides/agentops-visualization.png)

To leverage AgentOps, ensure you have your `AGENTOPS_API_KEY` configured in your `.env` file (as covered in the [Prerequisites](#prerequisites)). AgentOps automatically starts tracking your agent executions when you run `agentstack run` because of the `agentops.init()` call in `src/main.py`.

By using AgentOps, you move beyond simply building agents and gain powerful insights into their behavior and performance, allowing for continuous improvement and optimization.

> With a total run cost of only $0.01 in OpenAI credits (as seen in the AgentOps dashboard), this AI agent demonstrates exceptional cost-effectiveness. This was achieved entirely without code, eliminating the need for scraping, SQL queries, and other complex programming, paving the way for widespread application across various domains.

Reviewer: ```suggestion
With a total run cost of only $0.01 in OpenAI credits (as seen in the AgentOps dashboard), this AI agent runs efficiently while requiring no custom code. It works without scraping, SQL queries, or complex programming, making it accessible for a wide range of applications
```

Reviewer: toned down the hype a bit

</Admonition>

> **Congratulations!** You have successfully built and run a Web Scraper agent using AgentStack, Neon, and Firecrawl, demonstrating how to automate web data extraction and storage into a serverless Postgres database with minimal effort!

Reviewer: The rest of this guide needs to be tightened. I suggest going much shorter. Cut the summary section entirely and tighten Next Steps:

```
You’ve successfully built a Web Scraper agent using AgentStack, Neon, and Firecrawl, automating web data extraction and storage in a serverless Postgres database.

## Next Steps  

- **Explore more tools:** Browse AgentStack's [community tools](https://docs.agentstack.sh/tools/community), including Perplexity (knowledge retrieval), Composio (platform connections), and Stripe (payments).  
- **Expand your crew:** Add more agents and tasks to handle complex workflows.  
- **Monitor performance:** Use [AgentOps](https://docs.agentops.ai/) to track and optimize execution and costs.  

You can find the source code for the application described in this guide on GitHub.

```

- **Expand agent capabilities:** Experiment with adding more agents and tasks to create more complex and sophisticated workflows. Leverage the AgentStack CLI to easily generate and customize these components, building increasingly intelligent and autonomous systems.
- **Dive Deeper into Observability:** Explore AgentOps in more detail to monitor, benchmark, and optimize your agent's performance and costs. Gain valuable insights into your agent's execution and refine your applications for efficiency and effectiveness.

You can find the source code for the application described in this guide on GitHub.

<DetailIconCards>
    <a href="https://github.com/neondatabase-labs/neon-agenstack-example" description="AgentStack + Neon Example" icon="github">Building AI Agents with AgentStack and Neon</a>
</DetailIconCards>

## Resources

- [AgentStack Documentation](https://docs.agentstack.sh/introduction)
- [AgentOps Documentation](https://docs.agentops.ai/)
- [Firecrawl Documentation](https://docs.firecrawl.dev/introduction)
- [CrewAI Documentation](https://docs.crewai.com/introduction)

> - [Neon API Reference](https://api-docs.neon.tech/reference/getting-started-with-neon-api)

Reviewer: You could add community tools links for Firebase and Neon as well/instead:
https://docs.agentstack.sh/tools/tool/neon
https://docs.agentstack.sh/tools/tool/firecrawl
